### DISCUSSION TITLE
pinia store in library mode doesn't work

### DISCUSSION THREAD

**Discussion by aumaOne** [vuejs/pinia#2487] [Help and Questions] [answered]
### Reproduction

https://github.com/aumaOne/vite-pinia-library

### Steps to reproduce the bug

run `npm run build && npm link` on `my-lib`
run `npm link my-lib && npm run build && npm run preview` on `my-app`

### Expected behavior

build/preview would work the same as dev. 

### Actual behavior

All works fine when running with `npm run dev` but when using `npm run build && npm run preview` I'm getting a runtime error on the page
```
TypeError: Cannot read properties of undefined (reading '_s')
    at i (pinia.mjs:1714:20)
    at setup (HomeView.vue:5:22)
    at Tn (runtime-core.esm-bundler.js:158:18)
    at ag (runtime-core.esm-bundler.js:7290:25)
    at lg (runtime-core.esm-bundler.js:7251:36)
    at j (runtime-core.esm-bundler.js:5647:7)
    at O (runtime-core.esm-bundler.js:5613:9)
    at g (runtime-core.esm-bundler.js:5088:11)
    at ws.te [as fn] (runtime-core.esm-bundler.js:5821:9)
    at ws.run (reactivity.esm-bundler.js:178:19)
```

### Additional information

It seems that the lib can't access the pinia instance when it is built. pinia.mjs
```
...
pinia = activePinia;
        if (!pinia._s.has(id)) {
            // creating the store registers it in `pinia._s`
            if (isSetupStore) {
                createSetupStore(id, setup, options, pinia);
            }
            else {
                createOptionsStore(id, options, pinia);
            }
            /* istanbul ignore else */
            if ((process.env.NODE_ENV !== 'production')) {
                // @ts-expect-error: not the right inferred type
                useStore._pinia = pinia;
            }
        }
...
```

**Comment by 4color** [comment]
I also have such a problem here, have you solved it?

**Reply by aumaOne** [reply]
Unfortunately not. We tried different communities/channels and spend a lot of time to investigate, but decided to replace pinia in the library with a simple self implementation at the end.

**Comment by codelin-imu** [comment]
I encountered the same issue and resolved it by passing the active Pinia store to the defined store as param, in your case in the reproduction, to the `useCounterStore`. Pinia offers the [getActivePinia()](https://pinia.vuejs.org/api/modules/pinia.html#getActivePinia) function, which returns the currently active Pinia instance, if any. 

It seems that the problem may be related to how Vite resolves dependencies. Vite locates the nearest dependency to its root, and since Pinia is included in both the "my-app" and "my-lib" folders, it might be causing conflicts when importing a store from a other vite project / build. To troubleshoot this theory, consider removing Pinia from both projects and adding Pinia to the parent directory, placing it alongside "my-app" and "my-lib." as a shared dependency. Alternatively, consider adding an alias to both Vite configurations that targets the pinia dependency in the same node_modules folder.

**Reply by 4color** [reply]
I temporarily used this guy's method to handle it

https://github.com/vuejs/pinia/discussions/1073#discussioncomment-5279274

**Comment by Kuntey** [comment, marked as answer]
You can use vite's shared options [resolve.dedupe](https://vite.dev/config/shared-options.html#resolve-dedupe) to resolve this issue
```js
import { defineConfig } from 'vite'
export default defineConfig(() => {
  
  return {
    // ...
    resolve: {
      dedupe: ['pinia'],
    },
    // ...
  }
})
```

**Reply by dbaishya** [reply]
@Kuntey this is what did the trick for me. TY!

**Comment by slivos** [comment]
Maybe this helps. https://pinia.vuejs.org/core-concepts/outside-component-usage.html
